### ISSUE TITLE
[java] InsecureCryptoIv: False negative with fixed IVs from array initializers

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6531]
**Rule:** [InsecureCryptoIv](https://docs.pmd-code.org/latest/pmd_rules_java_security.html#insecurecryptoiv)

# False Negative: Fixed IVs are no longer tracked once they are moved into fields or helpers.

Version
7.22.0

## Checker
- Checker id: `category/java/security.xml/InsecureCryptoIv`
- Checker description: This rule detects insecure usage of hardcoded initialization vectors (IVs) in cryptographic operations by checking for direct instantiation of IvParameterSpec with constant arguments.

## Description of the false negative
`diff/output` marks 1 false negatives for `category/java/security.xml/InsecureCryptoIv`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The misses read more like matching gaps than deliberate exclusions.

### `PosCase1.java`
Source: `output/codeql/StaticInitializationVector/src/main/java/scensct/core/pos/PosCase1.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A method creates a byte array initialized with compile-time constants and passes it as IV to Cipher.init should be flagged as predictable IV.

```java
// A method creates a byte array initialized with compile-time constants and passes it as IV to Cipher.init should be flagged as predictable IV.
package scensct.core.pos;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.Key;

public class PosCase1 {
    public void encrypt() throws Exception {
        // Static byte array initialization with constants
        byte[] iv = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
        Key key = new SecretKeySpec(new byte[16], "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        // IV passed directly to Cipher.init
        cipher.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv)); // [REPORTED LINE]
    }
}
```

## Cause analysis
The rule is documented as follows: This rule detects insecure usage of hardcoded initialization vectors (IVs) in cryptographic operations by checking for direct instantiation of IvParameterSpec with constant arguments.

 The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/security.xml/InsecureCryptoIv`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by adangel** [comment]
@Carlson-JLQ Please use the issue templates we provide the next time. This ensures, that only one problem is reported per issue and it's easier to read. Thanks!

**Comment by adangel** [comment]
Note, that the original heading "Fixed IVs are no longer tracked once they are moved into fields or helpers" is wrong - the single test case doesn't show a problem with fields or helpers. It only uses a local variable. If there are different false negatives, please report them separately.
